Fix getTime field name to time in GetStats (#16894)

hye-on opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
Updated GetStats to deprecate the getTime field and introduce a new time field.

### Related Issues
Resolves #16894

### Comments
Since the API schema has been updated, I’ve prepared this PR to reflect the changes. [#785](https://github.com/opensearch-project/opensearch-api-specification/pull/785)
Being new to open source, I’m open to any suggestions or feedback. I appreciate your guidance and support! 

### Check List
- [x] Functionality includes testing.
  - Tested via curl commands:
    1. `curl -X PUT "localhost:9200/movies"`
    2. `curl "localhost:9200/movies/_stats?human&filter_path=_all.primaries.get`

<img width="1175" alt="image (1)" src="https://github.com/user-attachments/assets/769fb9e6-b763-4a19-ac60-d501632b9cd0" />

- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.   🌟 [#785](https://github.com/opensearch-project/opensearch-api-specification/pull/785)
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

reta commented on server/src/main/java/org/opensearch/index/get/GetStats.java:
```diff
@@ -155,7 +155,7 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         static final String GET = "get";
         static final String TOTAL = "total";
-        static final String TIME = "getTime";
+        static final String TIME = "time";
```
@hye-on thanks for this fix, since this is API change, changing the field name is a breaking change. We may need to expose this field as `getTime` (deprecated) and `time` (new one) for the time being.

hye-on replied:
@reta Thanks for the comment. I'll keep getTime as is and add time :)

reta commented on server/src/main/java/org/opensearch/index/get/GetStats.java:
```diff
@@ -155,6 +156,11 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+        /**
+         * Deprecated field name for time. Use {@link #TIME} instead.
+         */
+        @Deprecated
```
```suggestion
        @Deprecated(forRemoval = true)
```

reta commented on server/src/main/java/org/opensearch/index/get/GetStats.java:
```diff
@@ -137,6 +137,7 @@ public long current() {
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(Fields.GET);
         builder.field(Fields.TOTAL, getCount());
+        builder.humanReadableField(Fields.TIME_IN_MILLIS, Fields.GET_TIME, getTime());
```
👍

reta commented:
Last thing, could you please add a test case for it (see please https://github.com/opensearch-project/OpenSearch/tree/main/rest-api-spec/src/main/resources/rest-api-spec/test/indices.stats for existing tests), happy to help if you find it difficult. Thank you.

hye-on commented:
@reta Thank you for letting me know and offering your help! I’ll work on adding the test case and reach out if I run into any issues. I really appreciate your support. :)

hye-on commented:
@reta Thank you for your patience! I’ve added tests and updated the method handling the getTime field to address the `time_in_millis` duplication issue. Let me know if there’s anything else I should adjust :)

reta commented:
Thank you @hye-on , the build fails with:

```
* What went wrong:
Execution failed for task ':server:spotlessJavaCheck'.
> The following files had format violations:
      src/main/java/org/opensearch/index/get/GetStats.java
          @@ -138,7 +138,7 @@
           ····public·XContentBuilder·toXContent(XContentBuilder·builder,·Params·params)·throws·IOException·{
           ········builder.startObject(Fields.GET);
           ········builder.field(Fields.TOTAL,·getCount());
          -········builder.field(·Fields.GET_TIME,·Objects.toString(getTime()));
          +········builder.field(Fields.GET_TIME,·Objects.toString(getTime()));
           ········builder.humanReadableField(Fields.TIME_IN_MILLIS,·Fields.TIME,·getTime());
           ········builder.field(Fields.EXISTS_TOTAL,·existsCount);
           ········builder.humanReadableField(Fields.EXISTS_TIME_IN_MILLIS,·Fields.EXISTS_TIME,·getExistsTime());
  Run './gradlew :server:spotlessApply' to fix these violations.
```

hye-on commented:
> Thank you @hye-on , the build fails with:
> 

Apologies I missed that and thanks for pointing that out! [40ac728](https://github.com/opensearch-project/OpenSearch/pull/17009/commits/40ac7289acd4b09614ed92e291dfc75f1f29746b)

reta commented on rest-api-spec/src/main/resources/rest-api-spec/test/indices.stats/60_time_fields.yml.yml:
Seems like file name is a bit messed up: 60_time_fields.yml

hye-on replied:
@reta 
How about these options?

- 60_stats_time_field_migration.yml
- 60_stats_time_fields.yml
- 60_stats_time.yml

Thank you for your continued review :)

hye-on replied:
Thank you! I've renamed it to `60_include_both_time_and_gettime_in_stats.yml`

reta commented on rest-api-spec/src/main/resources/rest-api-spec/test/indices.stats/60_include_both_time_and_gettime_in_stats.yml:
```diff
@@ -0,0 +1,32 @@
+
+---
+"Test _stats API includes both time and getTime metrics with human filter":
+  - do:
```
@hye-on this is great, there are some tests failing, we need to add this section (just before `do` section) to exclude older OS versions (where this change is not present):

```
  - skip:
      version:  " - 2.19.99"
      reason:   "this change is added in 3.0.0"
```

reta commented on server/src/main/java/org/opensearch/index/get/GetStats.java:
```diff
@@ -137,6 +138,7 @@ public long current() {
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(Fields.GET);
         builder.field(Fields.TOTAL, getCount());
+        builder.field(Fields.GET_TIME, Objects.toString(getTime()));
```
@hye-on I am very sorry, missed that, I think we should use `humanReadableField` here, the same way the `Fields.TIME` was serialized before (no need to use `Objects.toString`):

```suggestion
        builder.humanReadableField(Fields.TIME_IN_MILLIS, Fields.GET_TIME, getTime());
```

hye-on replied:
@reta 
Initially, I tried using `humanReadableField`, but this method added both the first and second arguments, causing a duplicate error with `TIME_IN_MILLIS`.
To include only `getTime`, I checked the implementation of humanReadableField and found that it uses `Objects.toString(value)` for serialization.
For reference, calling `toString `on a `TimeValue` includes units like `s`or `ms`. 

#### Error details:
java.lang.RuntimeException: Failure at [indices.stats/60_time_fields.yml:11]: Duplicate field 'time_in_millis' at [Source: REDACTED (StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION disabled); line: 1, column: 150].

#### humanReadableField
```java
public XContentBuilder humanReadableField(String rawFieldName, String readableFieldName, Object value) throws IOException {
    assert rawFieldName.equals(readableFieldName) == false : "expected raw and readable field names to differ, but they were both: "
        + rawFieldName;
    if (humanReadable) {
        field(readableFieldName, Objects.toString(value)); // Adds the human-readable field using Objects.toString(value)
    }
    HumanReadableTransformer transformer = HUMAN_READABLE_TRANSFORMERS.get(value.getClass());
    if (transformer != null) {
        Object rawValue = transformer.rawValue(value);
        field(rawFieldName, rawValue); // Adds the raw field transformed 
    } else {
        throw new IllegalArgumentException("no raw transformer found for class " + value.getClass());
    }
    return this;
}

```

reta replied:
Ah, you are totally right, thank you

hye-on commented:
@reta 
Sorry, I've removed the unnecessary changelog entries. Please check it again. Thank you!
[2754fcb](https://github.com/opensearch-project/OpenSearch/pull/17009/commits/2754fcb3c6063bda5ac6090aa7754b59dc4054b8)

reta commented:
Thanks a lot @hye-on !
